### ISSUE TITLE
2nd rear camera problem on iPhone 13 with iOS 16.

### ISSUE THREAD

**Issue by ozknkrkc** [gruhn/vue-qrcode-reader#322]
When you read the qr code once or switch the front-back camera on the iPhone 13 with iOS 16, it switches to the second wide-angle camera and sees it as a side, and it works like this on the demo page, for example, I upload a video.

https://user-images.githubusercontent.com/16223128/216609011-4de38b24-f7b3-4984-bf61-a8456f1cd41d.mp4

**Comment by gruhn** [comment]
Are you on version v3.1.1 ?

**Comment by gruhn** [comment]
Alright, can you add the name of the bad camera to: 

https://github.com/gruhn/vue-qrcode-reader/blob/f7fd9cc812e2a6b3f3dfcff2b418b06b270fc374/src/misc/camera.js#L33

and open a pull request? Only make sure to follow the commit message convention. See:

https://github.com/gruhn/vue-qrcode-reader/blob/master/CONTRIBUTING.md#commit-messages

**Comment by gruhn** [comment]
Can you try adding other/multiple entries to `deviceBlackList` and make sure that your local setup is working. I won't be able to reproduce this, so I can't help you debug this.

**Comment by danbaechtold** [comment]
The "Desk View Camera" has different names in each language.
I added "de", "fr" and "it" in this PR: https://github.com/gruhn/vue-qrcode-reader/pull/326

@ozknkrkc [open this demo](https://gruhn.github.io/vue-qrcode-reader/select-camera-demo.html) to find the "bad" camera in your language.

**Comment by mstovicek** [comment]
oh, @danbaechtold , this may take a lot of time to fix for all users (listing camera names in different languages).

Isn't there another way to resolve this issue globally?

I'm going to (selfishly) add `Íróasztal nézet kamerája` (Hungarian camera name) to the blacklist.

**Comment by taco-c** [comment]
I used [patch-package](https://www.npmjs.com/package/patch-package) (you need to edit the file in the `dist` directory for it to work) to filter out both the desk view camera and the ultrawide camera in Norwegian.

I also read in #179 that the last device in the array is being selected as the default rear camera, but that seemingly does not work for iOS 16/iPhones. The order of the camera devices for me seem to be in the same order as the Turkish above:

1. Front camera
2. Rear camera
3. Ultrawide camera rear
4. Desk view camera

**Comment by Daltonmedia** [comment]
I tried to do a pull request for the Dutch translation. But I suppose I did something wrong, sorry...
Anway. Is someone kind enough to add "Bureauweergave" to the list? Thanks a lot.

**Comment by Daltonmedia** [comment]
I tried to do a pull request for the Dutch translation. But I suppose I did something wrong, sorry...
Anway. Is someone kind enough to add "Bureauweergave" to the list? Thanks a lot.

**Comment by alvinara23** [comment]
I added manually in the file: src/misc/camera.js the select camera for mi iphone device, its for spanish language, and dont work, any ideas?

![IMG_07BC83050D30-1](https://user-images.githubusercontent.com/18455334/228972192-c66b845f-13c8-4616-972d-0fbe75c993ce.jpeg)
![Captura de Pantalla 2023-03-30 a la(s) 15 50 54](https://user-images.githubusercontent.com/18455334/228972375-f6699723-601f-42b3-985f-3c1ce4b341ea.png)

**Comment by taco-c** [comment]
From testing with different versions of iOS 16 (16.1, 16.2, 16.3, 16.4), it seems only 16.3 has this problem. Updating to 16.4 (latest as of now) has fixed it for us in our webapp with no need for adding anything to the blacklist. The phones I tested on was in Norwegian and Finnish, whose camera names are not part of the already existing blacklist.

Could anyone else with this problem see if updating to iOS 16.4 fixes their problem?

**Comment by JKcolab** [comment]
> From testing with different versions of iOS 16 (16.1, 16.2, 16.3, 16.4), it seems only 16.3 has this problem. Updating to 16.4 (latest as of now) has fixed it for us in our webapp with no need for adding anything to the blacklist. The phones I tested on was in Norwegian and Finnish, whose camera names are not part of the already existing blacklist.
> 
> Could anyone else with this problem see if updating to iOS 16.4 fixes their problem?

Yes!(!!!!) 16.4 fixed it. Thank you!

**Comment by gruhn** [comment]
Not really fixed, but you can now use the `constraints` prop to set `deviceId` manually.

**Comment by heyFahad** [comment]
For anyone like me who's still on Vue 2 and can't upgrade to `vue-qrcode-reader@^5.0.0`, here’s a working patch that fixes the zoomed-in camera preview issue when reopening the scanner.

Since the newer release includes a fix but isn’t compatible with Vue 2, I used [patch-package](https://www.npmjs.com/package/patch-package) to backport the fix into the legacy version (`v3.2.1`) without upgrading.

## Summary of Changes

- Since I was only using the `QrcodeStream` component, I **patched** ONLY this `QrcodeStream` component to accept a new _optional_ `deviceId` prop, allowing my app to select a camera device on each open explicitly.

```diff
diff --git a/node_modules/vue-qrcode-reader/src/components/QrcodeStream.vue b/node_modules/vue-qrcode-reader/src/components/QrcodeStream.vue
index b482b5b..1a651a5 100644
--- a/node_modules/vue-qrcode-reader/src/components/QrcodeStream.vue
+++ b/node_modules/vue-qrcode-reader/src/components/QrcodeStream.vue
@@ -48,6 +48,11 @@ export default {
       }
     },
 
+    deviceId: {
+      type: String,
+      default: null
+    },
+
     torch: {
       type: Boolean,
       default: false
@@ -119,6 +124,12 @@ export default {
     },
 
     camera() {
+      console.log("Camera changed, re-initializing camera");
+      this.init();
+    },
+
+    deviceId() {
+      console.log("Device ID changed, re-initializing camera");
       this.init();
     }
   },
@@ -146,7 +157,8 @@ export default {
         } else {
           this.cameraInstance = await Camera(this.$refs.video, {
             camera: this.camera,
-            torch: this.torch
+            torch: this.torch,
+            deviceId: this.deviceId
           });
 
           const capabilities = this.cameraInstance.getCapabilities();
```

- Next, I updated my `misc/camera.js` file to **extend** the `MediaConstraints` object to respect the provided `deviceId`. If no `deviceId` is provided, the library continues to behave as originally intended.

```diff
diff --git a/node_modules/vue-qrcode-reader/src/misc/camera.js b/node_modules/vue-qrcode-reader/src/misc/camera.js
index 2d4f9fb..df181b6 100644
--- a/node_modules/vue-qrcode-reader/src/misc/camera.js
+++ b/node_modules/vue-qrcode-reader/src/misc/camera.js
@@ -48,6 +48,7 @@ const narrowDownFacingMode = async camera => {
     "Stolní kamera",
     "Ultrabrede camera aan voorzijde",
     "Front Ultra Wide Camera",
   ];
 
   const devices = (await navigator.mediaDevices.enumerateDevices())
@@ -60,7 +61,11 @@ const narrowDownFacingMode = async camera => {
     // devices for as the default front camera and the last entry as the default
     // rear camera seems to be a good heuristic on some devices.
     const frontCamera = devices[0];
-    const rearCamera = devices[devices.length - 1];
+    const defaultRearCamera = devices[devices.length - 1];
+    const preferredRearCamera = devices.find(d => /main|back camera|1x/i.test(d.label));
+
+    const rearCamera = preferredRearCamera || defaultRearCamera;
+    console.log({ devices, frontCamera, defaultRearCamera, preferredRearCamera, rearCamera });
 
     switch (camera) {
       case "auto":
@@ -86,7 +91,7 @@ const narrowDownFacingMode = async camera => {
   }
 };
 
-export default async function(videoEl, { camera, torch }) {
+export default async function(videoEl, { camera, torch, deviceId }) {
   // At least in Chrome `navigator.mediaDevices` is undefined when the page is
   // loaded using HTTP rather than HTTPS. Thus `STREAM_API_NOT_SUPPORTED` is
   // initialized with `false` although the API might actually be supported.
@@ -110,9 +115,11 @@ export default async function(videoEl, { camera, torch }) {
     video: {
       width: { min: 360, ideal: 640, max: 1920 },
       height: { min: 240, ideal: 480, max: 1080 },
-      ...(await narrowDownFacingMode(camera))
+      zoom: { ideal: 1 },
+      ...(deviceId ? { deviceId: { exact: deviceId } } : await narrowDownFacingMode(camera))
     }
   };
+  console.log({ camera, torch, deviceId, constraints });
 
   const stream = await navigator.mediaDevices.getUserMedia(constraints);
```

- Lastly, I **updated** the package.json entry point to `src/index.js` to avoid rebuilding this `vue-qrcode-reader` library manually under my deprecated Node 16.

```diff
diff --git a/node_modules/vue-qrcode-reader/package.json b/node_modules/vue-qrcode-reader/package.json
index d93c436..efa6554 100644
--- a/node_modules/vue-qrcode-reader/package.json
+++ b/node_modules/vue-qrcode-reader/package.json
@@ -9,7 +9,7 @@
   "repository": "github:gruhn/vue-qrcode-reader",
   "bugs": "https://github.com/gruhn/vue-qrcode-reader/issues",
   "homepage": "https://gruhn.github.io/vue-qrcode-reader",
-  "main": "dist/VueQrcodeReader.common.js",
+  "main": "src/index.js",
   "unpkg": "dist/VueQrcodeReader.umd.min.js",
   "keywords": [
     "vue",
```

- Finally, after applying the above changes, run the following command and commit the generated patch file to your repo so the fix persists across installs.

```
npx patch-package vue-qrcode-reader --exclude
```

This workaround has been stable in production for us and should help others running into the same issue on legacy stacks.

Let me know if this helps anyone in the future 👉🏼👈🏼
